From 8cdb4d978fa7c5294a5d1ee9365808549516623e Mon Sep 17 00:00:00 2001
From: Chin-Ting Kuo <chin-ting_kuo@aspeedtech.com>
Date: Sat, 9 Aug 2025 13:43:59 +0800
Subject: [PATCH] spi: aspeed: Porting set_speed callback function

Porting set_speed callback function in struct dm_spi_ops.
With this implementation, user can dynamically change SPI
clock frequency. However, user should be noticed that SPI
clock timing calibration mechanism is executed during
driver probe stage and the clock frequency adopted during
calibration process is the one configured in dts.

Signed-off-by: Chin-Ting Kuo <chin-ting_kuo@aspeedtech.com>
Change-Id: I469a85b45668d9fa1b5b0be396fdc506a70ffcd6

(cherry picked from commit 4e75bfeee8dc6393727f66f00e98b3d03c94f6a9)
Signed-off-by: Vasanthakumar K <vasanthakumark@ami.com>
---
 drivers/spi/spi-aspeed-smc.c | 81 +++++++++++++++++++-----------------
 1 file changed, 42 insertions(+), 39 deletions(-)

diff --git a/drivers/spi/spi-aspeed-smc.c b/drivers/spi/spi-aspeed-smc.c
index 8593856e6d..c31f8f62c0 100644
--- a/drivers/spi/spi-aspeed-smc.c
+++ b/drivers/spi/spi-aspeed-smc.c
@@ -101,7 +101,6 @@ struct aspeed_spi_flash {
 	uintptr_t ahb_base;
 	size_t ahb_decoded_sz;
 	u32 cmd_mode[CMD_MODE_MAX];
-	u32 max_freq;
 };
 
 struct aspeed_spi_priv {
@@ -127,7 +126,7 @@ struct aspeed_spi_info {
 	u32 (*segment_reg)(uintptr_t start, uintptr_t end);
 	bool (*decoded_range_lock)(struct udevice *bus);
 	int (*adjust_decoded_sz)(struct udevice *bus);
-	u32 (*get_clk_setting)(struct udevice *dev, uint hz);
+	u32 (*get_clk_setting)(struct udevice *bus, uint hz);
 	int (*calibrate)(struct udevice *dev, u32 hdiv,
 			 const u8 *golden_buf, u8 *test_buf);
 	void (*safs_init)(struct udevice *dev);
@@ -213,11 +212,9 @@ static void ast2400_spi_chip_set_4byte(struct udevice *bus, u32 cs)
 }
 
 /* Transfer maximum clock frequency to register setting */
-static u32 ast2400_get_clk_setting(struct udevice *dev, uint max_hz)
+static u32 ast2400_get_clk_setting(struct udevice *bus, uint max_hz)
 {
-	struct aspeed_spi_plat *plat = dev_get_plat(dev->parent);
-	struct aspeed_spi_priv *priv = dev_get_priv(dev->parent);
-	struct dm_spi_slave_plat *slave_plat = dev_get_parent_plat(dev);
+	struct aspeed_spi_plat *plat = dev_get_plat(bus);
 	u32 hclk_clk = plat->hclk_rate;
 	u32 hclk_div = 0x0000; /* default value */
 	u32 i;
@@ -234,17 +231,15 @@ static u32 ast2400_get_clk_setting(struct udevice *dev, uint max_hz)
 		}
 	}
 
-	if (found) {
+	if (found)
 		hclk_div = hclk_masks[i] << 8;
-		priv->flashes[slave_plat->cs].max_freq = hclk_clk / (i + 1);
-	}
 
-	dev_dbg(dev, "found: %s, hclk: %d, max_clk: %d\n", found ? "yes" : "no",
+	dev_dbg(bus, "found: %s, hclk: %d, max_clk: %d\n", found ? "yes" : "no",
 		hclk_clk, max_hz);
 
 	if (found) {
-		dev_dbg(dev, "h_div: %d (mask %x), speed: %d\n",
-			i + 1, hclk_masks[i], priv->flashes[slave_plat->cs].max_freq);
+		dev_dbg(bus, "h_div: %d (mask %x), speed: %d\n",
+			i + 1, hclk_masks[i], hclk_clk / (i + 1));
 	}
 
 	return hclk_div;
@@ -343,11 +338,9 @@ static int ast2500_adjust_decoded_size(struct udevice *bus)
 	return 0;
 }
 
-static u32 ast2500_get_clk_setting(struct udevice *dev, uint max_hz)
+static u32 ast2500_get_clk_setting(struct udevice *bus, uint max_hz)
 {
-	struct aspeed_spi_plat *plat = dev_get_plat(dev->parent);
-	struct aspeed_spi_priv *priv = dev_get_priv(dev->parent);
-	struct dm_spi_slave_plat *slave_plat = dev_get_parent_plat(dev);
+	struct aspeed_spi_plat *plat = dev_get_plat(bus);
 	u32 hclk_clk = plat->hclk_rate;
 	u32 hclk_div = 0x0000; /* default value */
 	u32 i;
@@ -360,8 +353,6 @@ static u32 ast2500_get_clk_setting(struct udevice *dev, uint max_hz)
 	for (i = 0; i < ARRAY_SIZE(hclk_masks); i++) {
 		if (hclk_clk / (i + 1) <= max_hz) {
 			found = true;
-			priv->flashes[slave_plat->cs].max_freq =
-							hclk_clk / (i + 1);
 			break;
 		}
 	}
@@ -374,8 +365,6 @@ static u32 ast2500_get_clk_setting(struct udevice *dev, uint max_hz)
 	for (i = 0; i < ARRAY_SIZE(hclk_masks); i++) {
 		if (hclk_clk / ((i + 1) * 4) <= max_hz) {
 			found = true;
-			priv->flashes[slave_plat->cs].max_freq =
-						hclk_clk / ((i + 1) * 4);
 			break;
 		}
 	}
@@ -384,12 +373,14 @@ static u32 ast2500_get_clk_setting(struct udevice *dev, uint max_hz)
 		hclk_div = BIT(13) | (hclk_masks[i] << 8);
 
 end:
-	dev_dbg(dev, "found: %s, hclk: %d, max_clk: %d\n", found ? "yes" : "no",
+	dev_dbg(bus, "found: %s, hclk: %d, max_clk: %d\n", found ? "yes" : "no",
 		hclk_clk, max_hz);
 
 	if (found) {
-		dev_dbg(dev, "h_div: %d (mask %x), speed: %d\n",
-			i + 1, hclk_masks[i], priv->flashes[slave_plat->cs].max_freq);
+		dev_dbg(bus, "h_div: %d (mask %x), speed: %d\n",
+			i + 1, hclk_masks[i],
+			!!(hclk_clk & BIT(13)) ? hclk_clk / ((i + 1) * 4) :
+			hclk_clk / (i + 1));
 	}
 
 	return hclk_div;
@@ -556,11 +547,9 @@ static int ast2700_adjust_decoded_size(struct udevice *bus)
 	return 0;
 }
 
-static u32 ast2600_get_clk_setting(struct udevice *dev, uint max_hz)
+static u32 ast2600_get_clk_setting(struct udevice *bus, uint max_hz)
 {
-	struct aspeed_spi_plat *plat = dev_get_plat(dev->parent);
-	struct aspeed_spi_priv *priv = dev_get_priv(dev->parent);
-	struct dm_spi_slave_plat *slave_plat = dev_get_parent_plat(dev);
+	struct aspeed_spi_plat *plat = dev_get_plat(bus);
 	u32 hclk_clk = plat->hclk_rate;
 	u32 hclk_div = 0x0400; /* default value */
 	u32 i, j;
@@ -584,26 +573,24 @@ static u32 ast2600_get_clk_setting(struct udevice *dev, uint max_hz)
 
 		if (found) {
 			hclk_div = ((j << 24) | hclk_masks[i] << 8);
-			priv->flashes[slave_plat->cs].max_freq =
-						hclk_clk / (i + 1 + j * 16);
 			break;
 		}
 	}
 
-	dev_dbg(dev, "found: %s, hclk: %d, max_clk: %d\n", found ? "yes" : "no",
+	dev_dbg(bus, "found: %s, hclk: %d, max_clk: %d\n", found ? "yes" : "no",
 		hclk_clk, max_hz);
 
 	if (found) {
-		dev_dbg(dev, "base_clk: %d, h_div: %d (mask %x), speed: %d\n",
-			j, i + 1, hclk_masks[i], priv->flashes[slave_plat->cs].max_freq);
+		dev_dbg(bus, "base_clk: %d, h_div: %d (mask %x), speed: %d\n",
+			j, i + 1, hclk_masks[i], hclk_clk / (i + 1 + j * 16));
 	}
 
 	return hclk_div;
 }
 
-static u32 ast2700_get_clk_setting(struct udevice *dev, uint max_hz)
+static u32 ast2700_get_clk_setting(struct udevice *bus, uint max_hz)
 {
-	return ast2600_get_clk_setting(dev, max_hz);
+	return ast2600_get_clk_setting(bus, max_hz);
 }
 
 void aspeed_spi_ast2700_safs_init(struct udevice *bus)
@@ -1042,10 +1029,9 @@ no_calib:
 	/* Nothing found ? */
 	if (best_div < 0) {
 		dev_warn(bus, "No good frequency\n");
-		clk_div = info->get_clk_setting(dev, max_freq);
+		clk_div = info->get_clk_setting(bus, max_freq);
 	} else {
 		dev_dbg(bus, "Found good read timings at HCLK/%d\n", best_div);
-		flash->max_freq = freq;
 		clk_div = ASPEED_SPI_HCLK_DIV(best_div - 1);
 	}
 
@@ -1325,7 +1311,7 @@ static int aspeed_spi_dirmap_create(struct spi_mem_dirmap_desc *desc)
 		/* assign SPI clock frequency division */
 		if (slave_plat->max_hz < (plat->hclk_rate / 5) ||
 		    priv->disable_calib) {
-			hclk_div = info->get_clk_setting(dev, slave_plat->max_hz);
+			hclk_div = info->get_clk_setting(bus, slave_plat->max_hz);
 
 			for (i = 0; i < CMD_MODE_MAX; i++) {
 				priv->flashes[i].cmd_mode[i] &=
@@ -1854,7 +1840,7 @@ static int aspeed_spi_claim_bus(struct udevice *dev)
 	dev_dbg(bus, "%s: claim bus CS%u\n", bus->name, slave_plat->cs);
 
 	if (slave_plat->max_hz < (plat->hclk_rate / 5) || priv->disable_calib) {
-		clk_setting = priv->info->get_clk_setting(dev, slave_plat->max_hz);
+		clk_setting = priv->info->get_clk_setting(bus, slave_plat->max_hz);
 		for (mode = 0; mode < CMD_MODE_MAX; mode++) {
 			flash->cmd_mode[mode] &= ~(priv->info->clk_ctrl_mask);
 			flash->cmd_mode[mode] |= clk_setting;
@@ -1886,13 +1872,30 @@ static int aspeed_spi_set_mode(struct udevice *bus, uint mode)
 
 static int aspeed_spi_set_speed(struct udevice *bus, uint hz)
 {
+	struct aspeed_spi_priv *priv = dev_get_priv(bus);
+	const struct aspeed_spi_info *info = priv->info;
+	struct aspeed_spi_regs *regs = priv->regs;
+	struct aspeed_spi_flash *flash;
+	enum aspeed_spi_cmd_mode mode;
+	u32 cs;
+
 	dev_dbg(bus, "%s: setting speed to %u\n", bus->name, hz);
 	/*
 	 * ASPEED SPI controller supports multiple CS with different
 	 * clock frequency. We cannot distinguish which CS here.
-	 * Thus, the related implementation is postponed to claim_bus.
+	 * Thus, the clock frequency of all CS will be changed once.
 	 */
 
+	for (cs = 0; cs < priv->num_cs; cs++) {
+		flash = &priv->flashes[cs];
+		for (mode = CMD_USER_MODE; mode < CMD_MODE_MAX; mode++) {
+			flash->cmd_mode[mode] &= ~info->clk_ctrl_mask;
+			flash->cmd_mode[mode] |= info->get_clk_setting(bus, hz);
+		}
+
+		writel(flash->cmd_mode[CMD_READ_MODE], &regs->ce_ctrl[cs]);
+	}
+
 	return 0;
 }
 
-- 
2.34.1

